generate_build_config.py

```diff
@@ -45,8 +46,48 @@
   permissions: '0755'
 """
 
+PRIVATE_PPA_TEMPLATE = """
+- chroot $CHROOT_ROOT apt install apt-transport-https
+- "echo 'deb {ppa_url} xenial main' | tee $CHROOT_ROOT/etc/apt/sources.list.d/builder-extra-ppa.list"
+- "chroot $CHROOT_ROOT apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys {key_id}"
+- chroot $CHROOT_ROOT apt update
+"""
+
+def _get_ppa_snippet(ppa, ppa_key=None):
+    """
+    Depending on what string is passed as PPA, return an appropriate yaml
+    snippet, ready to inject in TEMPLATE.
+
+    :param ppa:
+        The PPA URL. This should be either a "ppa:foo/bar" short form or a
+        full https:// URL for private PPAs.
+    :param ppa_key:
+        The hexacecimal key ID used to sign the PPA's package. This is only
+        used for private PPAs.
+    """
+    conf = ""
+    if ppa.startswith("https://"):
+        # This is likely a private PPA. We need to:
```

Comment:
It's not a great heuristic indeed, but my thinking was that user will either paste the right thing in, or paste in something stupid in which case apt-get update will blow up (and they get to keep both pieces).

I can refine that a little with a regexp, for instance checking that private-ppa.launchpad.net is in the URL or something, let me know if you think it would be better.